In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
%config InlineBackend.figure_format = 'retina'

## Q1

In [2]:
n_trials = nrows = 10_000
n_dice = ncols = 1

roll1 = np.random.choice([1, 2, 3, 4, 5, 6], n_trials * n_dice).reshape(nrows, ncols)
roll2 = np.random.choice([1, 2, 3, 4, 5, 6], n_trials * n_dice).reshape(nrows, ncols)

roll1
roll2

array([[6],
       [5],
       [2],
       ...,
       [1],
       [6],
       [3]])

In [58]:
rolls = np.random.choice([1,2,3,4,5,6], size = (100000, 2))
rolls

array([[1, 3],
       [2, 2],
       [2, 4],
       ...,
       [4, 3],
       [4, 4],
       [4, 6]])

In [59]:
(rolls[:,0] == rolls[:,1]).mean()

0.1661

In [3]:
doubles = roll1 == roll2
doubles

array([[ True],
       [False],
       [False],
       ...,
       [False],
       [False],
       [False]])

In [4]:
total_doubles = doubles.sum()
total_doubles

1673

In [5]:
doubles_rate = doubles.astype(int).mean()
doubles_rate

0.1673

The probability of rolling doubles when rolling two dice is .1609 out of 1, or ~16% of 100%

## Q2

In [6]:
n_trials = nrows = 10_000
n_coins = ncols = 8

flips = np.random.choice([0,1], n_trials * n_coins).reshape(nrows, ncols)
flips

array([[0, 1, 0, ..., 1, 1, 0],
       [0, 0, 0, ..., 1, 1, 0],
       [1, 1, 1, ..., 0, 1, 1],
       ...,
       [1, 0, 1, ..., 0, 1, 1],
       [0, 1, 1, ..., 0, 1, 0],
       [1, 1, 1, ..., 1, 1, 1]])

In [7]:
headcount = flips.sum(axis=1)
headcount

array([5, 3, 5, ..., 6, 3, 7])

In [8]:
success1 = headcount == 3
success1

array([False,  True, False, ..., False,  True, False])

In [9]:
success1.sum()

2216

In [10]:
success1_rate = success1.astype(int).mean()
success1_rate

0.2216

The probability of getting exactly 3 heads is ~22% of 100%

In [11]:
success2 = headcount > 3
success2

array([ True, False,  True, ...,  True, False,  True])

In [12]:
success2.sum()

6411

In [13]:
success2_rate = success2.astype(int).mean()
success2_rate

0.6411

The probability of getting more than 3 heads is ~64% of 100%

## Q3

In [14]:
n_trials = nrows = 10_000
n_alum = ncols = 2

alum = np.random.choice([1,2,3,4], n_trials * n_alum).reshape(nrows, ncols)
alum

array([[4, 2],
       [2, 3],
       [1, 4],
       ...,
       [3, 1],
       [2, 3],
       [3, 1]])

Let's say that a value of 1 means that the chosen student is from data science.

In [15]:
alum_select = alum.sum(axis=1)
alum_select

array([6, 5, 5, ..., 4, 5, 4])

In [16]:
double_ds = alum_select == 2
double_ds

array([False, False, False, ..., False, False, False])

In [17]:
double_ds.sum()

628

In [18]:
double_ds_rate = double_ds.astype(int).mean()
double_ds_rate

0.0628

There is a 6.24% probability of DS alumni being on two billboards passed in a row.

## Q4

In [60]:
n_trials = nrows = 10_000
n_pops = ncols = 5

pops_weekly=np.random.normal(3, 1.5, 50000).reshape(nrows,ncols)
pops_weekly

array([[-0.16266528,  4.01760181,  4.90879766,  1.57452652,  2.30087789],
       [ 4.31890948,  3.68805796,  2.30053775,  2.51828164,  2.10467431],
       [ 3.28761483,  2.00153407,  0.90936841, -0.58490341,  5.0223745 ],
       ...,
       [ 1.98145248,  5.08842063,  2.08140072,  1.69327044,  5.01752901],
       [ 3.27475418,  2.3005784 ,  4.82354352,  4.09105161,  2.69950988],
       [ 3.76753287,  1.83209075,  2.09330456,  3.19202689, -0.1766033 ]])

In [61]:
friday=17-pops_weekly.sum(axis=1)
friday

array([ 4.3608614 ,  2.06953884,  6.3640116 , ...,  1.13792671,
       -0.18943759,  6.29164823])

In [66]:
pop_fri = friday > 0
pop_fri

array([ True,  True,  True, ...,  True, False,  True])

In [67]:
pop_fri_rate = pop_fri.astype(int).mean()
pop_fri_rate

0.7183

There is a ~72% probability that there will be poptarts available by Friday afternoon.

## Q5

In [23]:
n_trials = nrows = 10_000
n_people = ncols = 2

men = np.random.normal(178, 8, nrows)
women = np.random.normal(170, 6, nrows)
#men_and_women = [men,women]
#men_and_women

In [24]:
women_taller = men < women
women_taller

array([False, False,  True, ..., False, False,  True])

In [25]:
women_taller.astype(int).mean()

0.2189

There is ~21% probability of the woman being taller than the man.

## Q6

In [73]:
n_trials = nrows = 10000
n_students = ncols = 50

installs = np.random.uniform(1,250, nrows*ncols).astype(int).reshape(nrows,ncols)
#installs

In [74]:
corrupt = installs == 1
#corrupt

In [75]:
corrupt_rate = corrupt.astype(int).mean()
corrupt_rate

0.004154

In [76]:
(corrupt.sum(axis=1) == 0).mean()

0.8137

Out of 50 students, ~.4% (1 student) had a corrupted download.

About 81% chance of installation resulting in no corruption.

In [93]:
n_trials = nrows = 10000
n_students = ncols = 100

installs = np.random.uniform(1,250, nrows*ncols).astype(int).reshape(nrows,ncols)
#installs

In [94]:
corrupt = installs == 1
#corrupt

In [95]:
corrupt_rate = corrupt.astype(int).mean()
corrupt_rate

0.004099

In [80]:
(corrupt.sum(axis=1) == 0).mean()

0.6627

~66% chance of no corruption

In [96]:
n_trials = nrows = 10000
n_students = ncols = 150

installs = np.random.uniform(1,250, nrows*ncols).astype(int).reshape(nrows,ncols)
#installs

In [97]:
corrupt = installs == 1

In [98]:
corrupt_rate = corrupt.astype(int).mean()
corrupt_rate

0.004036666666666667

In [99]:
(corrupt.sum(axis=1) == 0).mean()

0.5407

~54% chance of no corruption

In [100]:
n_trials = nrows = 10000
n_students = ncols = 450

installs = np.random.uniform(1,250, nrows*ncols).astype(int).reshape(nrows,ncols)
#installs

In [101]:
corrupt = installs == 1

In [102]:
corrupt_rate = corrupt.astype(int).mean()
corrupt_rate

0.004041111111111111

In [103]:
(corrupt.sum(axis=1) == 0).mean()

0.1668

~16.7% chance of no corruption

## Q7

In [104]:
n_trials = nrows = 10000
n_week = ncols = 7

truck_chance = np.random.random(nrows*ncols).reshape(nrows,ncols)
#truck_chance

In [105]:
truck_present = truck_chance <= .70
truck_present

array([[False,  True,  True, ...,  True, False,  True],
       [ True,  True,  True, ...,  True,  True,  True],
       [ True,  True,  True, ...,  True, False,  True],
       ...,
       [ True,  True,  True, ...,  True,  True,  True],
       [False,  True,  True, ...,  True,  True,  True],
       [ True,  True,  True, ...,  True, False,  True]])

In [106]:
truck_present_rate = truck_present.astype(int).mean()
truck_present_rate

0.6986714285714286

In [45]:
(truck_present.sum(axis=1) <= 4).mean()

0.3535

There is a 35% probability that a truck is present 4 or less days of the week.

In [46]:
(truck_present.sum(axis=1) >=1).mean()

0.9995

There is a 99.95% probability that a truck will be present this week.

## Q8

In [107]:
n_trials = nrows = 10000
n_people = ncols = 23

birthdays = np.random.uniform(1,366,nrows*ncols).astype(int).reshape(nrows,ncols)
#birthdays

In [110]:
n = pd.DataFrame(birthdays).nunique(axis = 1)

In [111]:
(n<23).mean()

0.5101